util/admission: TestWorkQueueBasic failed

util/admission.TestWorkQueueBasic [failed](https://tanzanite.cluster.engflow.com/invocations/default/26580ac7-4ec5-4724-ace1-ed3a6770bed0?testReportRun=1&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvdXRpbC9hZG1pc3Npb246YWRtaXNzaW9uX3Rlc3Q=) on release-24.1 @ [0cf3fc716870f66ea00844f88a030e6242afa8cf](https://github.com/cockroachdb/cockroach/commits/0cf3fc716870f66ea00844f88a030e6242afa8cf):


```
    datadriven.go:144: 
        /var/lib/engflow/worker/work/3/exec/bazel-out/k8-fastbuild/bin/pkg/util/admission/admission_test_/admission_test.runfiles/com_github_cockroachdb_cockroach/pkg/util/admission/testdata/work_queue:25:
        print [0 args]
        <no input to command>
        ----
        closed epoch: 0 tenantHeap len: 1 top tenant: 53
         tenant-id: 53 used: 1, w: 1, fifo: -128 waiting work heap: [0: pri: normal-pri, ct: 3, epoch: 0, qt: 100]
    datadriven.go:144: 
        /var/lib/engflow/worker/work/3/exec/bazel-out/k8-fastbuild/bin/pkg/util/admission/admission_test_/admission_test.runfiles/com_github_cockroachdb_cockroach/pkg/util/admission/testdata/work_queue:30:
        admit [5 args]
        <no input to command>
        ----
    datadriven.go:144: 
        /var/lib/engflow/worker/work/3/exec/bazel-out/k8-fastbuild/bin/pkg/util/admission/admission_test_/admission_test.runfiles/com_github_cockroachdb_cockroach/pkg/util/admission/testdata/work_queue:35:
        print [0 args]
        <no input to command>
        ----
        closed epoch: 0 tenantHeap len: 1 top tenant: 53
         tenant-id: 53 used: 1, w: 1, fifo: -128 waiting work heap: [0: pri: normal-pri, ct: 2, epoch: 0, qt: 100] [1: pri: normal-pri, ct: 3, epoch: 0, qt: 100]
    datadriven.go:144: 
        /var/lib/engflow/worker/work/3/exec/bazel-out/k8-fastbuild/bin/pkg/util/admission/admission_test_/admission_test.runfiles/com_github_cockroachdb_cockroach/pkg/util/admission/testdata/work_queue:41:
        admit [5 args]
        <no input to command>
        ----
    datadriven.go:144: 
        /var/lib/engflow/worker/work/3/exec/bazel-out/k8-fastbuild/bin/pkg/util/admission/admission_test_/admission_test.runfiles/com_github_cockroachdb_cockroach/pkg/util/admission/testdata/work_queue:46:
        admit [5 args]
        <no input to command>
        ----
    datadriven.go:144: 
        /var/lib/engflow/worker/work/3/exec/bazel-out/k8-fastbuild/bin/pkg/util/admission/admission_test_/admission_test.runfiles/com_github_cockroachdb_cockroach/pkg/util/admission/testdata/work_queue:50:
        print [0 args]
        <no input to command>
        ----
        closed epoch: 0 tenantHeap len: 2 top tenant: 71
         tenant-id: 53 used: 1, w: 1, fifo: -128 waiting work heap: [0: pri: normal-pri, ct: 2, epoch: 0, qt: 100] [1: pri: normal-pri, ct: 3, epoch: 0, qt: 100]
         tenant-id: 71 used: 0, w: 1, fifo: -128 waiting work heap: [0: pri: normal-pri, ct: 5, epoch: 0, qt: 100] [1: pri: low-pri, ct: 4, epoch: 0, qt: 100]
    datadriven.go:144: 
        /var/lib/engflow/worker/work/3/exec/bazel-out/k8-fastbuild/bin/pkg/util/admission/admission_test_/admission_test.runfiles/com_github_cockroachdb_cockroach/pkg/util/admission/testdata/work_queue:56:
         
        expected:
        continueGrantChain 5
        id 5: admit succeeded
        granted: returned 1
        
        found:
        continueGrantChain 5
        granted: returned 1
    panic.go:626: -- test log scope end --
--- FAIL: TestWorkQueueBasic (0.48s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=1</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/storage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestWorkQueueBasic.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-38926

--- a/pkg/util/admission/testdata/replicated_write_admission/tenant_fairness
+++ b/pkg/util/admission/testdata/replicated_write_admission/tenant_fairness
@@ -59,7 +59,7 @@ granter adjust-tokens=+0B
 print
 ----
 physical-stats: work-count=4 written-bytes=4B ingested-bytes=0B
-[regular work queue]: len(tenant-heap)=2 top-tenant=t2
+[regular work queue]: len(tenant-heap)=2 top-tenant=t1
  tenant=t1 weight=1 fifo-threshold=low-pri used=1B
   [0: pri=normal-pri create-time=1.002µs size=1B range=r1 origin=n1 log-position=4/21]
  tenant=t2 weight=1 fifo-threshold=low-pri used=1B

--- a/pkg/util/admission/testdata/replicated_write_admission/tenant_weights
+++ b/pkg/util/admission/testdata/replicated_write_admission/tenant_weights
@@ -49,7 +49,7 @@ admit tenant=t2 pri=normal-pri create-time=1.005us size=1B range=r2 origin=n1 lo
 print
 ----
 physical-stats: work-count=10 written-bytes=10B ingested-bytes=0B
-[regular work queue]: len(tenant-heap)=2 top-tenant=t1
+[regular work queue]: len(tenant-heap)=2 top-tenant=t2
  tenant=t1 weight=2 fifo-threshold=low-pri used=0B
   [0: pri=normal-pri create-time=1.001µs size=1B range=r1 origin=n1 log-position=4/20]
   [1: pri=normal-pri create-time=1.002µs size=1B range=r1 origin=n1 log-position=4/21]
@@ -75,8 +75,8 @@ granter class=regular adjust-tokens=+7B
 # towards the tenant with the higher weight (t2).
 grant class=regular
 ----
-admitted [tenant=t1 pri=normal-pri create-time=1.001µs size=1B range=r1 origin=n1 log-position=4/20]
 admitted [tenant=t2 pri=normal-pri create-time=1.001µs size=1B range=r2 origin=n1 log-position=5/20]
+admitted [tenant=t1 pri=normal-pri create-time=1.001µs size=1B range=r1 origin=n1 log-position=4/20]
 admitted [tenant=t2 pri=normal-pri create-time=1.002µs size=1B range=r2 origin=n1 log-position=5/21]
 admitted [tenant=t2 pri=normal-pri create-time=1.003µs size=1B range=r2 origin=n1 log-position=5/22]
 admitted [tenant=t1 pri=normal-pri create-time=1.002µs size=1B range=r1 origin=n1 log-position=4/21]

--- a/pkg/util/admission/testdata/work_queue
+++ b/pkg/util/admission/testdata/work_queue
@@ -63,7 +63,7 @@ granted: returned 1
 # favor of tenant 71.
 print
 ----
-closed epoch: 0 tenantHeap len: 2 top tenant: 71
+closed epoch: 0 tenantHeap len: 2 top tenant: 53
  tenant-id: 53 used: 1, w: 1, fifo: -128 waiting work heap: [0: pri: normal-pri, ct: 2, epoch: 0, qt: 100] [1: pri: normal-pri, ct: 3, epoch: 0, qt: 100]
  tenant-id: 71 used: 1, w: 1, fifo: -128 waiting work heap: [0: pri: low-pri, ct: 4, epoch: 0, qt: 100]
 
@@ -74,7 +74,7 @@ id 3: admit failed
 
 print
 ----
-closed epoch: 0 tenantHeap len: 2 top tenant: 71
+closed epoch: 0 tenantHeap len: 2 top tenant: 53
  tenant-id: 53 used: 1, w: 1, fifo: -128 waiting work heap: [0: pri: normal-pri, ct: 3, epoch: 0, qt: 100]
  tenant-id: 71 used: 1, w: 1, fifo: -128 waiting work heap: [0: pri: low-pri, ct: 4, epoch: 0, qt: 100]
 
@@ -551,20 +551,20 @@ admit id=2 tenant=10 priority=0 create-time-millis=1 bypass=false
 
 print
 ----
-closed epoch: 0 tenantHeap len: 2 top tenant: 5
+closed epoch: 0 tenantHeap len: 2 top tenant: 10
  tenant-id: 5 used: 0, w: 6, fifo: -128 waiting work heap: [0: pri: normal-pri, ct: 1, epoch: 0, qt: 100]
  tenant-id: 10 used: 0, w: 11, fifo: -128 waiting work heap: [0: pri: normal-pri, ct: 1, epoch: 0, qt: 100]
 
 granted chain-id=1
 ----
 continueGrantChain 1
-id 1: admit succeeded
+id 2: admit succeeded
 granted: returned 1
 
 granted chain-id=2
 ----
 continueGrantChain 2
-id 2: admit succeeded
+id 1: admit succeeded
 granted: returned 1
 
 # Both tenants are using 1 slot.
@@ -721,7 +721,7 @@ closed epoch: 0 tenantHeap len: 8 top tenant: 1
 # Set weights for 7 out of the 8 tenants.
 set-tenant-weights weights=1:2,2:3,3:4,4:5,5:6,7:8,8:9
 ----
-closed epoch: 0 tenantHeap len: 8 top tenant: 1
+closed epoch: 0 tenantHeap len: 8 top tenant: 8
  tenant-id: 1 used: 0, w: 2, fifo: -128 waiting work heap: [0: pri: normal-pri, ct: 1, epoch: 0, qt: 100]
  tenant-id: 2 used: 0, w: 3, fifo: -128 waiting work heap: [0: pri: normal-pri, ct: 1, epoch: 0, qt: 100]
  tenant-id: 3 used: 0, w: 4, fifo: -128 waiting work heap: [0: pri: normal-pri, ct: 1, epoch: 0, qt: 100]
@@ -734,49 +734,49 @@ closed epoch: 0 tenantHeap len: 8 top tenant: 1
 granted chain-id=1
 ----
 continueGrantChain 1
-id 1: admit succeeded
+id 8: admit succeeded
 granted: returned 1
 
 granted chain-id=2
 ----
 continueGrantChain 2
-id 8: admit succeeded
+id 7: admit succeeded
 granted: returned 1
 
 granted chain-id=3
 ----
 continueGrantChain 3
-id 7: admit succeeded
+id 5: admit succeeded
 granted: returned 1
 
 granted chain-id=4
 ----
 continueGrantChain 4
-id 6: admit succeeded
+id 4: admit succeeded
 granted: returned 1
 
 granted chain-id=5
 ----
 continueGrantChain 5
-id 5: admit succeeded
+id 3: admit succeeded
 granted: returned 1
 
 granted chain-id=6
 ----
 continueGrantChain 6
-id 4: admit succeeded
+id 2: admit succeeded
 granted: returned 1
 
 granted chain-id=7
 ----
 continueGrantChain 7
-id 3: admit succeeded
+id 1: admit succeeded
 granted: returned 1
 
 granted chain-id=8
 ----
 continueGrantChain 8
-id 2: admit succeeded
+id 6: admit succeeded
 granted: returned 1
 
 print

--- a/pkg/util/admission/work_queue.go
+++ b/pkg/util/admission/work_queue.go
@@ -15,6 +15,7 @@ import (
 	"context"
 	"fmt"
 	"math"
+	"slices"
 	"sort"
 	"sync"
 	"sync/atomic"
@@ -1012,27 +1013,33 @@ func (q *WorkQueue) SafeFormat(s redact.SafePrinter, _ rune) {
 		s.Printf("\n tenant-id: %d used: %d, w: %d, fifo: %d", tenant.id, tenant.used,
 			tenant.weight, tenant.fifoPriorityThreshold)
 		if len(tenant.waitingWorkHeap) > 0 {
+			// Sort items within waitingWorkHeap
+			sortedWaitingWorkHeap := slices.Clone(tenant.waitingWorkHeap)
+			sort.Sort(&sortedWaitingWorkHeap)
 			s.Printf(" waiting work heap:")
-			for i := range tenant.waitingWorkHeap {
+			for i := range sortedWaitingWorkHeap {
 				var workOrdering string
-				if tenant.waitingWorkHeap[i].arrivalTimeWorkOrdering == lifoWorkOrdering {
+				if sortedWaitingWorkHeap[i].arrivalTimeWorkOrdering == lifoWorkOrdering {
 					workOrdering = ", lifo-ordering"
 				}
 				s.Printf(" [%d: pri: %d, ct: %d, epoch: %d, qt: %d%s]", i,
-					tenant.waitingWorkHeap[i].priority,
-					tenant.waitingWorkHeap[i].createTime/int64(time.Millisecond),
-					tenant.waitingWorkHeap[i].epoch,
-					tenant.waitingWorkHeap[i].enqueueingTime.UnixNano()/int64(time.Millisecond), workOrdering)
+					sortedWaitingWorkHeap[i].priority,
+					sortedWaitingWorkHeap[i].createTime/int64(time.Millisecond),
+					sortedWaitingWorkHeap[i].epoch,
+					sortedWaitingWorkHeap[i].enqueueingTime.UnixNano()/int64(time.Millisecond), workOrdering)
 			}
 		}
 		if len(tenant.openEpochsHeap) > 0 {
+			// Sort items within openEpochsHeap
+			sortedOpenEpochsHeap := slices.Clone(tenant.openEpochsHeap)
+			sort.Sort(&sortedOpenEpochsHeap)
 			s.Printf(" open epochs heap:")
-			for i := range tenant.openEpochsHeap {
+			for i := range sortedOpenEpochsHeap {
 				s.Printf(" [%d: pri: %d, ct: %d, epoch: %d, qt: %d]", i,
-					tenant.openEpochsHeap[i].priority,
-					tenant.openEpochsHeap[i].createTime/int64(time.Millisecond),
-					tenant.openEpochsHeap[i].epoch,
-					tenant.openEpochsHeap[i].enqueueingTime.UnixNano()/int64(time.Millisecond))
+					sortedOpenEpochsHeap[i].priority,
+					sortedOpenEpochsHeap[i].createTime/int64(time.Millisecond),
+					sortedOpenEpochsHeap[i].epoch,
+					sortedOpenEpochsHeap[i].enqueueingTime.UnixNano()/int64(time.Millisecond))
 			}
 		}
 	}
@@ -1403,7 +1410,15 @@ func (th *tenantHeap) Len() int {
 }
 
 func (th *tenantHeap) Less(i, j int) bool {
-	// used_i/weight_i < used_j/weight_j
+	// For tenant fairness, use used_i/weight_i < used_j/weight_j to determine
+	// order. In case of a tie, prioritize items with higher weight, and then
+	// items with lower tenant id.
+	if (*th)[i].used*uint64((*th)[j].weight) == (*th)[j].used*uint64((*th)[i].weight) {
+		if (*th)[i].weight == (*th)[j].weight {
+			return (*th)[i].id < (*th)[j].id
+		}
+		return (*th)[i].weight > (*th)[j].weight
+	}
 	return (*th)[i].used*uint64((*th)[j].weight) < (*th)[j].used*uint64((*th)[i].weight)
 }
 

--- a/pkg/util/admission/work_queue_test.go
+++ b/pkg/util/admission/work_queue_test.go
@@ -48,6 +48,12 @@ func (b *builderWithMu) printf(format string, a ...interface{}) {
 	fmt.Fprintf(&b.buf, format, a...)
 }
 
+func (b *builderWithMu) String() string {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	return b.buf.String()
+}
+
 func (b *builderWithMu) stringAndReset() string {
 	b.mu.Lock()
 	defer b.mu.Unlock()
@@ -88,18 +94,6 @@ func (tg *testGranter) tookWithoutPermission(count int64) {
 
 func (tg *testGranter) continueGrantChain(grantChainID grantChainID) {
 	tg.buf.printf("continueGrantChain%s %d", tg.name, grantChainID)
-}
-
-func (tg *testGranter) grant(grantChainID grantChainID) {
-	rv := tg.r.granted(grantChainID)
-	if rv > 0 {
-		// Need deterministic output, and this is racing with the goroutine that
-		// was admitted. Sleep to let it get scheduled. We could do something more
-		// sophisticated like monitoring goroutine states like in
-		// concurrency_manager_test.go.
-		time.Sleep(150 * time.Millisecond)
-	}
-	tg.buf.printf("granted%s: returned %d", tg.name, rv)
 }
 
 func (tg *testGranter) storeWriteDone(
@@ -248,8 +242,8 @@ func TestWorkQueueBasic(t *testing.T) {
 					}
 				}(ctx, workInfo, id)
 				// Need deterministic output, and this is racing with the goroutine
-				// which is trying to get admitted. Sleep to let it get scheduled.
-				time.Sleep(50 * time.Millisecond)
+				// which is trying to get admitted. Retry to let it get scheduled.
+				maybeRetryWithWait(t, d.Expected, d.Rewrite, buf.String)
 				return buf.stringAndReset()
 
 			case "set-try-get-return-value":
@@ -261,7 +255,13 @@ func TestWorkQueueBasic(t *testing.T) {
 			case "granted":
 				var chainID int
 				d.ScanArgs(t, "chain-id", &chainID)
-				tg.grant(grantChainID(chainID))
+				rv := tg.r.granted(grantChainID(chainID))
+				if rv > 0 {
+					// Need deterministic output, and this is racing with the goroutine that was
+					// admitted. Retry a few times.
+					maybeRetryWithWait(t, d.Expected, d.Rewrite, buf.String)
+				}
+				tg.buf.printf("granted%s: returned %d", tg.name, rv)
 				return buf.stringAndReset()
 
 			case "cancel-work":
@@ -276,8 +276,8 @@ func TestWorkQueueBasic(t *testing.T) {
 				}
 				work.cancel()
 				// Need deterministic output, and this is racing with the goroutine
-				// whose work is canceled. Sleep to let it get scheduled.
-				time.Sleep(50 * time.Millisecond)
+				// whose work is canceled. Retry to let it get scheduled.
+				maybeRetryWithWait(t, d.Expected, d.Rewrite, buf.String)
 				return buf.stringAndReset()
 
 			case "work-done":
@@ -319,6 +319,11 @@ func TestWorkQueueBasic(t *testing.T) {
 				return q.String()
 
 			case "print":
+				// Need deterministic output, and this is racing with the goroutine
+				// whose work is canceled. Retry to let it get scheduled.
+				maybeRetryWithWait(t, d.Expected, d.Rewrite, func() string {
+					return q.String()
+				})
 				return q.String()
 
 			case "advance-time":
@@ -343,6 +348,26 @@ func scanTenantID(t *testing.T, d *datadriven.TestData) roachpb.TenantID {
 	var id int
 	d.ScanArgs(t, "tenant", &id)
 	return roachpb.MustMakeTenantID(uint64(id))
+}
+
+func maybeRetryWithWait(t *testing.T, expected string, rewrite bool, f func() string) {
+	t.Helper()
+	if rewrite {
+		time.Sleep(100 * time.Millisecond)
+		return
+	}
+	var str string
+	// Retry for at most 2s (20 iterations with 100ms delay).
+	for i := 0; i < 20; i++ {
+		str = f()
+		if strings.TrimSpace(str) == strings.TrimSpace(expected) {
+			if i > 0 {
+				t.Logf("retried test case %d times", i)
+			}
+			return
+		}
+		time.Sleep(100 * time.Millisecond)
+	}
 }
 
 // TestWorkQueueTokenResetRace induces racing between tenantInfo.used
@@ -382,6 +407,7 @@ func TestWorkQueueTokenResetRace(t *testing.T) {
 						TenantID:   w.tenantID,
 						CreateTime: createTime,
 					})
+					buf.printf("admit")
 					require.Equal(t, true, enabled)
 					mu.Lock()
 					defer mu.Unlock()
@@ -392,7 +418,12 @@ func TestWorkQueueTokenResetRace(t *testing.T) {
 				}(ctx, work2, createTime)
 				createTime++
 				if work != nil {
-					tg.grant(1)
+					rv := tg.r.granted(1)
+					if rv > 0 {
+						// Need deterministic output, and this is racing with the goroutine that was
+						// admitted. Add a retry to let it get scheduled.
+						maybeRetryWithWait(t, "admit", false /* rewrite */, buf.String)
+					}
 					work.cancel()
 					buf.stringAndReset()
 				}
@@ -402,7 +433,13 @@ func TestWorkQueueTokenResetRace(t *testing.T) {
 			}
 			if work != nil {
 				work.cancel()
-				tg.grant(1)
+				rv := tg.r.granted(1)
+				if rv > 0 {
+					// Need deterministic output, and this is racing with the goroutine that was
+					// admitted. Add a retry to let it get scheduled.
+					maybeRetryWithWait(t, "admit", false /* rewrite */, buf.String)
+				}
+				buf.stringAndReset()
 			}
 		}
 	}()
@@ -540,6 +577,7 @@ func TestStoreWorkQueueBasic(t *testing.T) {
 				opts.usesTokens = true
 				opts.timeSource = timeutil.NewManualTime(timeutil.FromUnixMicros(0))
 				opts.disableEpochClosingGoroutine = true
+				opts.disableGCTenantsAndResetUsed = true
 				st = cluster.MakeTestingClusterSettings()
 				var mockCoordMu syncutil.Mutex
 				q = makeStoreWorkQueue(log.MakeTestingAmbientContext(tracing.NewTracer()), roachpb.StoreID(1),
@@ -586,8 +624,8 @@ func TestStoreWorkQueueBasic(t *testing.T) {
 					}
 				}(ctx, workInfo, id)
 				// Need deterministic output, and this is racing with the goroutine
-				// which is trying to get admitted. Sleep to let it get scheduled.
-				time.Sleep(50 * time.Millisecond)
+				// which is trying to get admitted. Retry to let it get scheduled.
+				maybeRetryWithWait(t, d.Expected, d.Rewrite, buf.String)
 				return buf.stringAndReset()
 
 			case "set-try-get-return-value":
@@ -607,7 +645,13 @@ func TestStoreWorkQueueBasic(t *testing.T) {
 
 			case "granted":
 				wc := tryScanWorkClass(t, d)
-				tg[wc].grant(0)
+				rv := tg[wc].r.granted(0)
+				if rv > 0 {
+					// Need deterministic output, and this is racing with the goroutine that was
+					// admitted. Retry a few times.
+					maybeRetryWithWait(t, d.Expected, d.Rewrite, buf.String)
+				}
+				buf.printf("granted%s: returned %d", tg[wc].name, rv)
 				return buf.stringAndReset()
 
 			case "cancel-work":
@@ -622,8 +666,8 @@ func TestStoreWorkQueueBasic(t *testing.T) {
 				}
 				work.cancel()
 				// Need deterministic output, and this is racing with the goroutine
-				// whose work is canceled. Sleep to let it get scheduled.
-				time.Sleep(50 * time.Millisecond)
+				// whose work is canceled. Retry to let it get scheduled.
+				maybeRetryWithWait(t, d.Expected, d.Rewrite, buf.String)
 				return buf.stringAndReset()
 
 			case "work-done":
@@ -679,6 +723,11 @@ func TestStoreWorkQueueBasic(t *testing.T) {
 				return printQueue()
 
 			case "print":
+				// Need deterministic output, and this is racing with the goroutine
+				// whose work is canceled. Retry to let it get scheduled.
+				maybeRetryWithWait(t, d.Expected, d.Rewrite, func() string {
+					return printQueue()
+				})
 				return printQueue()
 
 			default:

